### PR TITLE
#166 - support non-opengl 2D mode

### PR DESCRIPTION
**Pull request by wijnen**
Here's the pull request for #166 as you requested.  The patch applied to the latest version, so that was easy.  I originally wrote the "canvas" needed to be a div; that's no longer true.  This version expects it to be a canvas, making it easier to select the mode from javascript without needing any changes in the html.

### CODE REVIEW

File: src/m.builder.js [hunk 1]
```diff
@@ -130,6 +130,8 @@ please.builder.SpriteBuilder.prototype = {
 
     // builds and returns a VBO
     "build" : function () {
+        if (please.gl.ctx === null)
+            return {};
```
Comment (Aeva) [inline src/m.builder.js]: please use { braces } to denote the block, for clarity

File: src/m.camera.js [hunk 2]
```diff
@@ -83,7 +83,7 @@ please.CameraNode = function () {
     this.__is_camera = true;
 
     please.make_animatable_tripple(this, "look_at", "xyz", [0, 0, 0]);
-    please.make_animatable_tripple(this, "up_vector", "xyz", [0, 0, 1]);
+    please.make_animatable_tripple(this, "up_vector", "xyz", please.gl.ctx === null ? [0, 1, 0] : [0, 0, 1]);
```
Comment (Aeva) [inline src/m.camera.js]: In this case, we never want this property to be able to change, correct?  The conditional should instead be that please.make_animatable_tripple is only ever called when please.gl.ctx is defined, otherwise it should be set to [0, 0, 1] and have the property frozen or something.

File: src/m.camera.js [hunk 3]
```diff
@@ -105,9 +105,12 @@ please.CameraNode = function () {
     ANI("near", 0.1);
     ANI("far", 100.0);
 
+    if (please.gl.ctx === null)
+        this.location_z = 100.0;
```
Comment (Aeva) [inline src/m.camera.js]: please use { braces } to denote the block, for clarity

File: src/m.camera.js [hunk 4]
```diff
@@ -105,9 +105,12 @@ please.CameraNode = function () {
...
+
     this.mark_dirty();
     this.projection_matrix = mat4.create();
-    this.__projection_mode = "perspective";
+    this.__projection_mode = please.gl.ctx === null ? "orthographic" : "perspective";
```
Comment (Aeva) [inline src/m.camera.js]: Another place where the property should never change, so a conditional statement following the original definition which sets it to orthographic and then calls Object.freeze(this.__projection_mode);

File: src/m.gl.js [hunk 5]
```diff
@@ -47,6 +47,11 @@ please.gl.set_context = function (canvas_id, options) {
 
     this.canvas = document.getElementById(canvas_id);
     please.__create_canvas_overlay();
+
+    if (options == '2d') {
+        return;
```
Comment (Aeva) [inline src/m.gl.js]: I think please.gl.ctx being null should not indicate that we're in 2D mode, but rather something else should indicate that its in 2D mode.  The reason for this, is if someone calls stuff before set_context, then there should be an entirely different behavior (possibly throwing an error) than assuming 2D or 3D mode, because the user never explicitly stated what they needed.

File: src/m.graph.js [hunk 6]
```diff
@@ -407,7 +407,8 @@ please.GraphNode = function () {
 
     // GLSL bindings with default driver methods:
     this.__regen_glsl_bindings = function (event) {
-        var prog = please.gl.__cache.current;
+        if (please.gl.ctx !== null)
+            var prog = please.gl.__cache.current;
```
Comment (Aeva) [inline src/m.graph.js]: please use { braces } to denote the block, for clarity

File: src/m.graph.js [hunk 7]
```diff
@@ -432,11 +433,13 @@ please.GraphNode = function () {
         please.make_animatable(
             this, "billboard_mode", this.__billboard_driver, this.shader, true);
 
-        // prog.samplers is a subset of prog.vars
-        for (var name, i=0; i<prog.uniform_list.length; i+=1) {
-            name = prog.uniform_list[i];
-            if (ignore.indexOf(name) === -1 && !this.shader.hasOwnProperty(name)) {
-                please.make_animatable(this, name, null, this.shader);
+        if (please.gl.ctx !== null) {
```
Comment (Aeva) [inline src/m.graph.js]: I'm unsure if behavior that is specific to 2D mode should be inside a function called "regen_glsl_bindings".  Maybe some of the functionality in this function should be folded out into another method.

File: src/m.graph.js [hunk 8]
```diff
@@ -669,6 +672,10 @@ please.GraphNode.prototype = {
         // binds, so as long as the calls are sorted, this extra
         // overhead should be insignificant.
         if (this.visible && this.__drawable && typeof(this.draw) === "function") {
+            if (please.gl.ctx === null) {
```
Comment (Aeva) [inline src/m.graph.js]: the graph's drawing code should never be called at all in 2D mode.  This just introduces an extra branch into something that is called every frame for 3D mode to prevent some behavior in 2D mode.

File: src/m.graph.js [hunk 9]
```diff
@@ -854,6 +861,8 @@ please.SceneGraph = function () {
         var prog = please.gl.get_program();
         if (this.camera) {
             this.camera.update_camera();
+            if (please.gl.ctx === null)
```
Comment (Aeva) [inline src/m.graph.js]: please use { braces } to denote the block, for clarity

File: src/m.graph.js [hunk 10]
```diff
@@ -1062,6 +1071,8 @@ please.pipeline.add(-1, "mgrl/picking_pass", function () {
 // Picking RenderNode
 //
 please.SceneGraph.prototype.__create_picking_node = function () {
+    if (please.gl.ctx === null)
+        return null;
```
Comment (Aeva) [inline src/m.graph.js]: the picking stuff should never be called in 2D mode.

File: src/m.media.js [hunk 11]
```diff
@@ -574,7 +574,6 @@ please.media.__image_instance = function (center, scale, x, y, width, height, al
     node.ibo = data.ibo;
     node.ext = {};
     node.vars = {};
-    node.shader["diffuse_texture"] = this.asset_name,
```
Comment (Aeva) [inline src/m.media.js]: why was this removed?

File: src/m.media.js [hunk 12]
```diff
@@ -583,7 +582,13 @@ please.media.__image_instance = function (center, scale, x, y, width, height, al
...
     node.draw_type = "sprite";
     node.sort_mode = "alpha";
-
+    if (please.gl.ctx === null) {
+        var div = please.overlay.new_element();
+        div.appendChild(this);
+        div.bind_to_node(node);
+        return node;
+    }
+    node.shader["diffuse_texture"] = this.asset_name,
```
Comment (Aeva) [inline src/m.media.js]: should probably be in an else clause

**Comment by wijnen** [comment]
I'm preparing a new version for you to review.  Do you want me to squash the commits so it remains a single commit to m.grl, or should I keep the history as separate commits?

**Comment by wijnen** [comment]
Never mind; it's too much trouble merging without squashing, so one commit will have to do.

**Comment by wijnen** [comment]
Ok, I've updated the patch.  It's quite a bit longer now.  Here's what I did:
- Possibly very important, because I'm not sure if it breaks gl rendering: I moved node.shader.world_matrix to node.world_matrix, because I needed it in 2-D mode, but node.shader is not defined there.
- Add m.dom.js containing 2-D specific parts.
- Add please.renderer which is please.dom or please.gl, depending on which was initialized.  It is an object of mostly function pointers for doing renderer-specific operations.  They are called from everywhere.
- Add a function sync() to please.SceneGraph, which should be called for 2-D operations instead of draw().  The draw() function calls it before doing a lot of gl things.
- Move some global picker code into a function that is only called for the gl renderer.

Things that probably still need to be done (I didn't test, but didn't implement them):
- Scaling of image nodes.
- Gani animations; or do they work automatically?
- Z stacking.  I'm not sure about the best way to do this.  The z-index CSS property is ignored for position:absolute elements.

**Comment by Aeva** [comment]
Getting close!  I need to find some time to review more in depth, but the one thing that stands out at the moment is that you pulled a number of areas out of files like m.graph.js and m.camera.js and put them into m.gl.js.  I would say that pulling that functionality out into its own method like you did is the right call, but it should still live in the same file.  Eg, code that pertains to the camera should live in m.camera.js.  m.gl.js exists entirely to provide dynamic bindings to shaders and abstract the crunchier parts of webgl like VBOs.

If you like, I can help clean up this feature (I think it'll mostly just be minor changes) so that we can merge it expediently, but I'd need commit access to your fork.

**Comment by Aeva** [comment]
also, moving the world_matrix stuff out of node.shader will likely break a lot of things.  I'm fine the "shader" namespace on graph nodes still being a thing (sans complicated glsl bindings of course) on non-gl mode for now.

**Comment by wijnen** [comment]
Of course you may help.  You can make any changes after you merge it anyway. ;-)  I've added you to my fork, so you should be able to push to it.

I'm fine with putting no function implementations in the please.renderer definitions; they are part of both the file they are used in and the renderer implementation, so they make sense in either file.

I considered making an almost empty shader object for the dom renderer.  It seems to me like the shader is really a gl thing, so it's out of place there, but it's not a big deal.  Also, I think performance should be a higher priority, so if using a shader object helps for that, that's certainly a good reason to go that way.

**Comment by Aeva** [comment]
working on the PR will be helpful because I can use github to isolate easily on what has been currently changed, so that way when I pull this into master, it shouldn't break anything.

the node.shader thing serves two purposes, the first being to provide a convenient namespace for data bindings that the developer will probably not need to use frequently, and the second being to provide a place for automatic data bindings to glsl uniform vars, which are them selves data bindings.  The world_matrix is there not as a client side convinience, but because glsl needs it to do anything, and it corresponds to a shader binding of the same name.

At this point, I'm willing to accept something called "shader" being where utility data bindings are defined as a misnomer, because none of this code was ever intended to run without gl, and so is very tightly coupled to it

**Comment by wijnen** [comment]
I just force-pushed a commit that removes generated files from the patch, so you can expect some warnings when pulling; I hope this makes it easier to rebase if you change master.  I'll work on moving world_matrix back into the shader object now.

**Comment by Aeva** [comment]
@wijnen I just merged a very large feature branch to master btw (glsl compiler); so I can help get this PR in shape for merging soon.

**Comment by Aeva** [comment]
merge conflict is resolved, so you might want to verify that your changes still work

**Comment by wijnen** [comment]
THe test program I'm running (http://wijnen.dtdns.net/~shevek/2d-test/) isn't
very extensive, but it still works fine.

**Comment by Aeva** [comment]
Hey - I refactored the branch in preparation for merging it.  Some changes:
 - please.dom.set_frame is now please.dom.set_context
 - please.renderer an object defined in m.defs.js, which both set_context functions modify
 - please.renderer.width and please.renderer.height are now getters that return what you'd expect
 - removed the renderer.allow_picking thing since it is gl specific, and made it so the machinery is only added when using the gl renderer.  In the future, I think it would not be hard to add picking functionality with a compatible api for the 2D renderer.
 - I added a really pdq "nogl" demo so I can test things... given the output, I think its likely that I broke that functionality.

So, with the above in mind, could test this build and let me know if your stuff works?
Also, I think there might be a performance regression in the webgl renderer that I need to investigate.

Overall, I'm pretty happy with this, and I think we can merge it soon.  Sorry for delay in getting this together.


Also - I think I want to change how images are instanced in the DOM renderer, so that they are just a div with a background set, instead of an image inside of a div object.  This will be faster to render, and it gives us a lot more control via css.

**Comment by wijnen** [comment]
I'm trying a more complex example (I didn't look at yours yet, perhaps it has the same issue). orthographic_grid seems to be ignored, and the full size of the overlay is always [-1:1] on both axes. I briefly tried to find the cause of this, but didn't find it. Perhaps some update code isn't run periodically (graph.sync is run; I checked).

I was afraid it might hurt the OpenGL performance. Since this is an extra feature, IMO that should be fixed before this can be merged.

**Comment by wijnen** [comment]
The placement problem seems to be in the overlay code, not in this pull request. I'll open a separate issue for it.
